acq4/devices/PatchPipette/states.py

```diff
@@ -1130,97 +1161,257 @@ class PatchPipetteResealState(PatchPipetteState):
+
+    def processAtLeastOneTestPulse(self):
+        """Wait for at least one test pulse to be processed."""
+        while True:
+            self.checkStop()
+            tps = self.getTestPulses(timeout=0.2)
+            if len(tps) > 0:
+                break
+            self.sleep(0.2)
+        tps = [(tp.startTime(), tp.analysis()) for tp in tps]
+        self._rollingRecentTestPulses.extend(tps)
+
+        last_second = tps[-1][0] - 1
+        while self._rollingRecentTestPulses[0][0] < last_second:
+            self._rollingRecentTestPulses.popleft()
+        self.analyzeRecentTestPulses()
+
+        self._rollingLongTermTestPulses.extend(tps)
+        last_20_seconds = tps[-1][0] - 20
+        while self._rollingLongTermTestPulses[0][0] < last_20_seconds:
+            self._rollingLongTermTestPulses.popleft()
+        if self._updateRollingResistanceThresholds:
+            self.analyzeLongTermTestPulses()
+
+    def analyzeLongTermTestPulses(self):
+        too_recent = self._rollingLongTermTestPulses[-1][0] - 1
+        access_res = np.fromiter(
+            (a[1]['peakResistance']
+             for a in self._rollingLongTermTestPulses if a[0] < too_recent),
+            dtype=float,
+        )
+        input_res = np.fromiter(
+            (a[1]['steadyStateResistance'] - a[1]['peakResistance']
+             for a in self._rollingLongTermTestPulses if a[0] < too_recent),
+            dtype=float,
+        )
+        # TODO this may need to be optimized not to recalculate the mean and variance every time
+        self._longTermInputResistanceMean = np.mean(input_res)
+        self._longTermInputResistanceVariance = np.var(input_res)
+        self._longTermAccessResistanceMean = np.mean(access_res)
+        self._longTermAccessResistanceVariance = np.var(access_res)
+
+    def analyzeRecentTestPulses(self):
+        analysis = [tp[1] for tp in self._rollingRecentTestPulses]
+        access_res = np.fromiter((a['peakResistance'] for a in analysis), dtype=float)
+        input_res = np.fromiter((a['steadyStateResistance'] - a['peakResistance'] for a in analysis), dtype=float)
+        self._recentInputResistanceMean = np.mean(input_res)
+        self._recentInputResistanceVariance = np.var(input_res)
+        self._recentAccessResistanceMean = np.mean(access_res)
+        self._recentAccessResistanceVariance = np.var(access_res)
+
+    def freezeRollingResistanceThresholds(self):
+        """Stop updating the rolling averages"""
+        self._updateRollingResistanceThresholds = False
+
+    def unfreezeRollingResistanceThresholds(self):
+        self._updateRollingResistanceThresholds = True
 
     def run(self):
         config = self.config
         dev = self.dev
-        self.monitorTestPulse()
+        if config['extractNucleus'] is True:
+            self.sealToNucleus()
+        dev.pressureDevice.setPressure(config['initialPressureSource'], config['initialPressure'])
```

Comment:
Why this? Shouldn't sealToNucleus leave us at the desired pressure? (and if it isn't called, then we are already at the initial pressure anyway)

Reply:
I'd assume the procedure looks like:
- nuzzle nucleus while slowly ramping up vacuum
- hold vacuum constant and slowly retract
- watch for changes in resistance